### PR TITLE
add Iterators.countfrom(::Any, ::Any)

### LINKED ISSUE #37462: `Iterators.countfrom(::Any, ::Any)` doesn't exist
Is there a reason why only `Iterators.countfrom(::Number, ::Number)` is defined ?

I wanted to do 
```julia
using Dates
Iterators.countfrom(now(), Minute(15))
```

I know I can do the same with `Number`s but I like this syntax
It doesn't look like it would be an issue to allow it:
https://github.com/JuliaLang/julia/blob/697e782ab86bfcdd7fd15550241fe162c51d9f98/base/iterators.jl#L597

### PR DESCRIPTION
**Pull request by MarcMush**
closes #37462

### CODE REVIEW

**Review by JeffBezanson — APPROVED** [review]
Gotta love ordinal types :)

**Comment by MarcMush** [comment]
@JeffBezanson can this be added for 1.6 ?

**Comment by StefanKarpinski** [comment]
Bump?

**Comment by stevengj** [comment]
Needs a NEWS item?

**Comment by MarcMush** [comment]
sorry for the delay, I added a NEWS entry

File: test/iterators.jl [hunk 1]
```diff
@@ -133,6 +133,12 @@ let i = 0, k = 1
         k += 1
         k <= 10 || break
     end
+    for j = countfrom(Int[0,0], Float64[1.0,2.0])
+        @test j isa Vector{Float64}
+        @test j == l*[1,2]
+        l += 1
+        l <= 10 || break
+    end
```
Comment (stevengj) [inline test/iterators.jl:141]: Maybe have a test case for types where T and S are different, e.g.
```jl
@test collect(take(countfrom(Dates.Date("2020-12-25"), Dates.Day(1)), 12)) == range(Dates.Date("2020-12-25"), step=Dates.Day(1), length=12)
```

Comment (MarcMush) [reply test/iterators.jl:142]: indeed, here T is promoted to `Vector{Float64}` as well
is it ok to use `Dates` in tests or should I use types from Base Julia ?

File: NEWS.md [hunk 2]
```diff
@@ -51,6 +51,7 @@ Standard library changes
   constructing the range. ([#40382])
 * TCP socket objects now expose `closewrite` functionality and support half-open mode usage ([#40783]).
 * Intersect returns a result with the eltype of the type-promoted eltypes of the two inputs ([#41769]).
+* `Iterators.countfrom` now accepts any type that define `+`. ([#37747])
```
Comment (stevengj) [inline NEWS.md:54]:
```suggestion
* `Iterators.countfrom` now accepts any type that defines `+`. ([#37747])
```

**Review by stevengj — APPROVED** [review]

File: base/iterators.jl [hunk 3]
```diff
@@ -613,11 +613,13 @@ julia> for v in Iterators.countfrom(5, 2)
...
-countfrom(start::Number)               = Count(start, oneunit(start))
-countfrom()                            = Count(1, 1)
+countfrom(start::T, step::S) where {T,S} = Count{typeof(start+step),S}(start, step)
+countfrom(start::Number, step::Number)   = Count(promote(start, step)...)
+countfrom(start)                         = Count(start, oneunit(start))
+countfrom()                              = Count(1, 1)
 
-eltype(::Type{Count{S}}) where {S} = S
+
+eltype(::Type{Count{T,S}}) where {T,S} = T
```
Comment (vtjnash) [inline base/iterators.jl:622]:
```suggestion
eltype(::Type{<:Count{T}}) where {T} = T
```

**Comment by oscardssmith** [comment]
I accidentally just merged this. Does it need to be reverted @vtjnash?

**Pull request merged**